cot/src/common_types.rs

```diff
@@ -132,6 +132,233 @@ impl From<String> for Password {
         Self::new(password)
     }
 }
+/// A validated URL wrapper.
+///
+/// This structure ensures that the contained URL is correctly formatted and
+/// complies with standard URL syntax rules. It wraps [`url::Url`] to provide
+/// validation upon construction through methods like [`new`] and [`from_str`],
+/// and exposes useful methods for accessing URL components or converting the
+/// URL into different representations.
+///
+/// # Examples
+///
+/// ## Creating a Validated URL Using `new`
+///
+/// ```
+/// use cot::common_types::Url;
+///
+/// // Successful URL creation
+/// let url = Url::new("https://example.com").unwrap();
+///
+/// // Accessing the normalized URL string
+/// assert_eq!(url.as_str(), "https://example.com/");
+/// ```
+///
+/// ## Parsing a URL from a String Using `from_str`
+///
+/// ```
+/// use std::str::FromStr;
+///
+/// use cot::common_types::Url;
+///
+/// // Parse a valid URL string
+/// let url = Url::from_str("https://example.com").unwrap();
+///
+/// // Convert into owned string representation
+/// let url_string = url.into_string();
+/// assert_eq!(url_string, "https://example.com/");
+/// ```
+///
+/// # Behavior
+///
+/// - **Validation**: Both `new` and `from_str` ensure the input is a syntactically
+///   correct URL as defined by the WHATWG URL specification via the underlying
+///   [`url::Url`] parser.
+/// - **Normalization**: The internal URL is normalized (e.g., trailing slash added
+///   for HTTP URLs) during construction.
+#[derive(Clone, Debug)]
```

Comment:
```suggestion
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```

These might prove to be useful for users.